src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -216,45 +216,54 @@ internal sealed partial class VisualStudioProject
+                }
 
-            field = newValue;
+                var oldValue = field;
 
-            // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
-            // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
-            // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
-            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
-            // fully loaded.
-            var isFullyLoadedTask = _workspaceStatusService?.IsFullyLoadedAsync(CancellationToken.None);
-            var isFullyLoaded = isFullyLoadedTask is { IsCompleted: true } && isFullyLoadedTask.GetAwaiter().GetResult();
+                field = newValue;
 
-            // We only log telemetry during solution open
-            if (logThrowAwayTelemetry && _telemetryService?.HasActiveSession == true && !isFullyLoaded)
-            {
-                TryReportCompilationThrownAway(_workspace.CurrentSolution.State, Id);
-            }
+                // Importantly, we do not await/wait on the fullyLoadedStateTask.  We do not want to ever be waiting on work
+                // that may end up touching the UI thread (As we can deadlock if GetTagsSynchronous waits on us).  Instead,
+                // we only check if the Task is completed.  Prior to that we will assume we are still loading.  Once this
+                // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
+                // fully loaded.
+                var isFullyLoadedTask = _workspaceStatusService?.IsFullyLoadedAsync(CancellationToken.None);
+                var isFullyLoaded = isFullyLoadedTask is { IsCompleted: true } && isFullyLoadedTask.GetAwaiter().GetResult();
 
-            if (_activeBatchScopes > 0)
-            {
-                _projectPropertyModificationsInBatch.Add(withNewValue);
-            }
-            else
-            {
-                _workspace.ApplyChangeToWorkspace(Id, withNewValue);
+                // We only log telemetry during solution open
+                if (logThrowAwayTelemetry && _telemetryService?.HasActiveSession == true && !isFullyLoaded)
+                {
+                    TryReportCompilationThrownAway(_workspace.CurrentSolution.State, Id);
+                }
+
+                if (_activeBatchScopes > 0)
+                {
+                    _projectPropertyModificationsInBatch.Add(solutionChanges => updateSolution(solutionChanges, oldValue));
+                }
+                else
+                {
+                    _workspace.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync: false, solutionChanges =>
+                    {
+                        updateSolution(solutionChanges, oldValue);
+                    }).AsTask().GetAwaiter().GetResult();
```

Comment:
consider doc'ing tha tthis is ok since we pass in useAsync:false.

Reply:
Oh right I need to just make a ApplyBatchChange helper that does that.

Reply:
💭 Need to make a `CompletedResult()` helper that throws when the operation is not already complete at the time of the call.

Reply:
Added a method called "AssertComplete" and used it in the appropriate places. To keep things a bit simpler, I just added an ApplyBatchChangeToWorkspace() function that uses that new helper.